### ISSUE TITLE
[Feature Request] Extensibility for POP with SHR

### ISSUE THREAD

**Issue by bgavrilMS** [AzureAD/microsoft-authentication-library-for-go#439]
Request for MSAL to enable extensibility for POP + SHR. This is a request from Kubernetes team.

Proposed interface for extensibility: 

```csharp
    interface IAuthenticationScheme
    {
        /// <summary>
        /// Extra parameters that are added to the request to the /token endpoint. 
        /// For POP this would be req_cnf=(public key) and token_type=pop
        /// </summary>
        IReadOnlyDictionary<string, string> GetTokenRequestParams();

        /// <summary>
        /// Key ID to which to associate the ESTS token response with. Used by MSAL for caching puroposes.
        /// </summary>
        string KeyId { get; }

        /// <summary>
        /// Creates the access token that goes into an Authorization HTTP header. For POP this means adding the SHR claims and signing 
        /// the token with the client key. 
        /// </summary>
        string FormatAccessToken(string accessToken); 
    }
```
See [POP implementation](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/AuthScheme/PoP/PoPAuthenticationScheme.cs) for this interface in MSAL.NET, which does in-memory key management as well.

MSAL changes: 

- [ ] Add new API to all requests in MSAL GO, e.g. `AcquireTokenForClient(...).WithAuthenticationScheme(popScheme)`
- [ ] Add the parameters from `GetTokenRequestParams` to the /token request as POST data
- [ ] When token response comes back from ESTS, associate the access token to the `KeyId`. Make sure this association is reflected in the cache (see below)
- [ ] Before constructing the `AuthenticationResult`, format the access token by invoking `FormatAccessToken` (this will create the actual POP token. Remember that MSAL doesn't cache the formatted POP token!)
- [ ] Update token cache lookup logic to take into account `keyId`
- [ ] If token is found in cache and matches `keyId`, make sure to call `FormatAccessToken` before constructing `AuthenticationResult`

Changes to the cache

- [ ] Add -{authschme} to the access token key. This is to differentiate between bearer and pop tokens. Only 1 pop token will be supported. If this part of the key is not present, assume the token is "Bearer". 
- [ ] Add key to the payload to associate the token with the kid.
- [ ] Ensure token type is "pop" (AAD returns the token_type)

Example pop access token in cache:

```json
{
  "AccessToken": {
    "-login.windows.net-accesstoken_with_authscheme-16dab2ba-145d-4b1b-8569-bf4b9aed4dc8-72f988bf-86f1-41af-91ab-2d7cd011db47-https://vault.azure.net/.default-pop": {
      "environment": "login.windows.net",
      "client_id": "16dab2ba-145d-4b1b-8569-bf4b9aed4dc8",
      "secret": "actual_token",
      "credential_type": "AccessToken",
      "realm": "72f988bf-86f1-41af-91ab-2d7cd011db47",
      "target": "https://vault.azure.net/.default",
      "cached_at": "1688149697",
      "expires_on": "1688236096",
      "extended_expires_on": "1688236096",
      "kid": "TJtUe-3Wa8X6cOxKzmeTxjHDO1-gXBKtazWh0xx3SA0",    # new!
      "token_type": "pop",   # not Bearer!
      "refresh_on": "1688192896",
      "ext_expires_on": "1688236096"
    }
  }
``` 

Example ESTS response: 

<img width="166" alt="image" src="https://github.com/AzureAD/microsoft-authentication-library-for-go/assets/12273384/d5d797d9-4c71-491b-8438-83f2afb3dd6a">

**Comment by bgavrilMS** [comment]
CC @jmprieur  and @localden

**Comment by localden** [comment]
@bgavrilMS makes sense to expose this through extensibility. In terms of claims, when making the request do we need to do anything special here, similar to what the [Node.js folks did](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/shr-client-claims.md)?

**Comment by bgavrilMS** [comment]
Agreed, the SHR is tied to the request, not to the app.

**Comment by julienstroheker** [comment]
PR is been tracked here #443

**Comment by bgavrilMS** [comment]
Fixed with #443

**Comment by kevin-xu-glean** [comment]
Hi all, Azure noob herer! 

We are looking to support PoP OAuth authentication with the go library here. Is that possible today or do we need to wait for this feature request to be fully implemented?
